jupyter-docker/jupyter_localize_extension.py

```diff
@@ -1,19 +1,57 @@
-from notebook.utils import url_path_join
+import subprocess
+import os
+import pipes
+import tornado
+from tornado import gen
+from tornado.web import HTTPError
 from notebook.base.handlers import IPythonHandler
+from notebook.utils import url_path_join
 
 class LocalizeHandler(IPythonHandler):
-    def post(self):
-        self.set_status(200)
-        self.finish()
+  def sanitize(self, pathstr):
+    """Expands to absolute paths, makes intermediate dirs, and quotes to remove any shell naughtiness.
+    This doesn't need to be a coroutine as it's an inline function, not something Future-y."""
+    #expanduser behaves fine with gs:// urls, thankfully
+    expanded = os.path.expanduser(pathstr)
+    if not pathstr.startswith("gs://"):
+      try:
+        os.makedirs(expanded)
+      except OSError: #thrown if dirs already exist
+        pass
+    return pipes.quote(expanded)
+
+  @gen.coroutine
+  def localize(self, pathdict):
+    """Treats the given dict as a string/string map and sends it to gsutil."""
+    #This gets dropped inside the user's notebook working directory
+    with open("localization.log", 'a') as locout:
+      for key in pathdict:
+        cmd = " ".join(["gsutil -m -q cp -R", self.sanitize(key), self.sanitize(pathdict[key])])
```

Comment:
Q: should ` { foo : bar }` put `foo` in `bar` or `bar` in `foo`?

Option 1: foo = dest, bar = source. Allows writing the same source file to multiple destinations.  
Option 2:  foo = source, bar = dest. Makes sense in English but doesn't let you write the same source to multiple destinations.  
Option 3: foo = source, bar = array[dest]. Makes the code more complicated (have to handle both array and single-value RHSes) but makes sense in English AND is flexible.